src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) 2012-2016 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "wallet/coinselection.h"
+#include "util.h"
+#include "utilmoneystr.h"
+
+// Descending order comparator
+struct {
+    bool operator()(CInputCoin a, CInputCoin b) const
+    {
+        return a.txout.nValue > b.txout.nValue;
+    }
+} descending;
+
+struct CompareValueOnly
+{
+    bool operator()(const CInputCoin& t1,
+                    const CInputCoin& t2) const
+    {
+        return t1.txout.nValue < t2.txout.nValue;
+    }
+};
+
+bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret, FastRandomContext* rand)
+{
+    out_set.clear();
+    value_ret = 0;
+
+    if (utxo_pool.size() <=0) {
+        return false;
+    }
+
+    int depth = 0;
+    int tries = 100000;
+    std::vector<std::pair<bool, bool>> selection; // First bool: select the utxo at this index; Second bool: traversing second branch of this utxo
+    selection.assign(utxo_pool.size(), std::pair<bool, bool>(false, false));
+    bool done = false;
+    bool backtrack = false;
+    
+    // Sort the utxo_pool
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+    
+    // Calculate remaining
+    CAmount remaining = 0;
+    for (CInputCoin utxo : utxo_pool) {
+        remaining += utxo.txout.nValue;
```

Comment:
@xekyo we should assume that it would be terrible unless someone can show that it will not cause another massive UTXO bloat event... but thats offtopic here, as I don't think anyone has this concern with exact matches.